File: src/test/java/shopping/application/member/MemberServiceTest.java
package shopping.application.member;

import jakarta.transaction.Transactional;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.security.crypto.password.PasswordEncoder;
import shopping.domain.member.Member;
import shopping.domain.member.exception.LoginFailedException;
import shopping.domain.repository.MemberRepository;
import shopping.dto.LoginRequest;
import shopping.dto.LoginResponse;
import shopping.dto.SignUpRequest;
import shopping.infrastructure.auth.JwtTokenProvider;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

@SpringBootTest
@Transactional
class MemberServiceTest {
    @Autowired
    private MemberService memberService;

    @Autowired
    private MemberRepository memberRepository;

    @Autowired
    private PasswordEncoder passwordEncoder;

    @Autowired
    private JwtTokenProvider jwtTokenProvider;

    @Test
    @DisplayName("회원 가입에 성공하고, 비밀번호는 암호화되어 저장되어야 한다.")
    void signUp_success_and_password_encrypted() {
        String rawPassword = "password123!";
        SignUpRequest request = new SignUpRequest("dev@example.com", rawPassword);

        // when
        Long savedId = memberService.signUp(request);

        // then
        Member foundMember = memberRepository.findById(savedId)
                .orElseThrow(() -> new AssertionError("회원이 정상적으로 저장되지 않았습니다."));

        assertThat(foundMember.getEmail()).isEqualTo("dev@example.com");
        assertThat(foundMember.getPassword()).isNotEqualTo(rawPassword);
        assertThat(passwordEncoder.matches(rawPassword, foundMember.getPassword())).isTrue();
    }

    @Test
    @DisplayName("이메일과 비밀번호로 맞으면 JWT 토큰이 발급되어야 한다.")
    void login_success() {
        String email = "login@example.com";
        String password = "password123!";
        memberService.signUp(new SignUpRequest(email, password));

        LoginResponse response = memberService.login(new LoginRequest(email, password));

        // then
        assertThat(response.accessToken()).isNotBlank();

        // 토큰이 유효한지 provider를 통해 한 번 더 검증 (선택)
        assertThat(jwtTokenProvider.validateToken(response.accessToken())).isTrue();
        assertThat(jwtTokenProvider.getPayload(response.accessToken())).isEqualTo(email);
    }

    @Test
    @DisplayName("비밀번호가 틀리면 로그인에 실패하고 예외가 발생해야 한다.")
    void login_fail_with_wrong_password() {
        // given
        String email = "wrong_pw@example.com";
        memberService.signUp(new SignUpRequest(email, "correct_password"));

        // when & then
        assertThatThrownBy(() -> memberService.login(new LoginRequest(email, "wrong_password")))
                .isInstanceOf(LoginFailedException.class);
    }
}

File: src/test/kotlin/shopping/application/member/MemberServiceTest.kt
package shopping.application.member

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk
import org.instancio.Instancio
import org.instancio.Select.field
import shopping.domain.member.Member
import shopping.domain.member.MemberRepository
import shopping.support.error.CoreException

class MemberServiceTest : BehaviorSpec({

    val memberRepository = mockk<MemberRepository>()
    val memberService = MemberService(memberRepository)

    given("회원가입 시") {

        `when`("이메일이 중복되지 않으면") {
            val member = Instancio.of(Member::class.java)
                .set(field("email"), "test@test.com")
                .set(field("password"), "password123")
                .create()

            every { memberRepository.existsByEmail(member.email) } returns false
            every { memberRepository.save(any()) } returns member

            then("회원가입에 성공한다") {
                val result = memberService.register(member.email, member.password)
                result.email shouldBe member.email
            }
        }

        `when`("이메일이 이미 존재하면") {
            val member = Instancio.create(Member::class.java)

            every { memberRepository.existsByEmail(member.email) } returns true

            then("예외가 발생한다") {
                val exception = shouldThrow<CoreException> {
                    memberService.register(member.email, member.password)
                }
                exception.message shouldBe "사용자가 이미 존재합니다."
            }
        }
    }

    given("로그인 시") {

        `when`("이메일과 비밀번호가 일치하면") {
            val member = Instancio.of(Member::class.java)
                .set(field("email"), "test@test.com")
                .set(field("password"), "password123")
                .create()

            every { memberRepository.findByEmail(member.email) } returns member

            then("멤버 ID를 반환한다") {
                val result = memberService.login(member.email, member.password)
                result shouldBe member.id
            }
        }

        `when`("이메일이 존재하지 않으면") {
            val member = Instancio.create(Member::class.java)

            every { memberRepository.findByEmail(member.email) } returns null

            then("예외가 발생한다") {
                val exception = shouldThrow<CoreException> {
                    memberService.login(member.email, member.password)
                }
                exception.message shouldBe "사용자 존재하지 않습니다."
            }
        }

        `when`("비밀번호가 틀리면") {
            val member = Instancio.of(Member::class.java)
                .set(field("email"), "test@test.com")
                .set(field("password"), "correctPassword")
                .create()

            every { memberRepository.findByEmail(member.email) } returns member

            then("예외가 발생한다") {
                val exception = shouldThrow<CoreException> {
                    memberService.login(member.email, "wrongPassword")
                }
                exception.message shouldBe "비밀번호가 틀렸습니다."
            }
        }
    }
})
